database/database.go

```diff
@@ -486,3 +489,75 @@ func (s *DatabaseService) DeleteExecutionPayloads(idFirst, idLast uint64) error
 	_, err := s.DB.Exec(query, idFirst, idLast)
 	return err
 }
+
+func (s *DatabaseService) UpsertBuilderDemotion(submitBlockRequest *types.BuilderSubmitBlockRequest, signedBlindedBeaconBlock *types.SignedBlindedBeaconBlock, signedValidatorRegistration *types.SignedValidatorRegistration) error {
+	if submitBlockRequest == nil {
+		return fmt.Errorf("nil submitBlockRequest invalid for UpsertBuilderDemotion")
+	}
+
+	_submitBlockRequest, err := json.Marshal(submitBlockRequest)
+	if err != nil {
+		return err
+	}
+	bidTrace := submitBlockRequest.Message
+	builderDemotionEntry := BuilderDemotionEntry{
+		SubmitBlockRequest: NewNullString(string(_submitBlockRequest)),
+
+		Slot:  bidTrace.Slot,
+		Epoch: bidTrace.Slot / uint64(common.SlotsPerEpoch),
+
+		BuilderPubkey:  bidTrace.BuilderPubkey.String(),
+		ProposerPubkey: bidTrace.ProposerPubkey.String(),
+
+		Value: bidTrace.Value.String(),
+
+		BlockHash: bidTrace.BlockHash.String(),
+	}
+
+	if signedBlindedBeaconBlock != nil {
+		_signedBlindedBeaconBlock, err := json.Marshal(signedBlindedBeaconBlock)
+		if err != nil {
+			return err
+		}
+		builderDemotionEntry.SignedBlindedBeaconBlock = NewNullString(string(_signedBlindedBeaconBlock))
+	}
+
+	if signedValidatorRegistration != nil {
```

Comment:
You made a good point that we don't really need the validator registration in the demotion because the fee recipient and gas limit are part of the prechecks. This should remove a bunch of lines!

Reply:
Gah, I may have messed up here. We do need the validator registration to know which fee recipient to refund. Having the full signed validator registration will leave a cryptographic trace for our records.